### PR TITLE
Fixturize tests/frame/test_dtypes.py

### PR DESCRIPTION
**Pull request by h-vetinari**
One more steps towards #22471. Again, needing to re-add some fixtures that were removed in #24885.

### CODE REVIEW

File: pandas/tests/frame/conftest.py [hunk 1]
```diff
@@ -110,6 +110,18 @@ def mixed_int_frame():
     return df
 
 
+@pytest.fixture
+def mixed_type_frame():
```
Comment (jreback) [inline pandas/tests/frame/conftest.py]: so this fixtures itself is prob ok here. as we *know* that this is used in many places.

File: pandas/tests/frame/test_dtypes.py [hunk 2]
```diff
@@ -410,41 +409,41 @@ def test_dtypes_gh8722(self):
...
                                D='float64:dense')).sort_values()
         result = frame.ftypes.sort_values()
         assert_series_equal(result, expected)
 
-    def test_astype(self):
-        casted = self.frame.astype(int)
-        expected = DataFrame(self.frame.values.astype(int),
-                             index=self.frame.index,
-                             columns=self.frame.columns)
+    def test_astype(self, float_frame, mixed_float_frame, mixed_type_frame):
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: see my comments elsewhere. don't use this pattern.

File: pandas/tests/frame/test_dtypes.py [hunk 3]
```diff
@@ -473,37 +474,37 @@ def _check_cast(df, v):
...
+        df = float_frame.copy()
+        expected = float_frame.astype(np.int32)
         df['string'] = 'foo'
         casted = df.astype(np.int32, errors='ignore')
 
         expected['string'] = 'foo'
         assert_frame_equal(casted, expected)
 
-    def test_astype_with_view(self):
+    def test_astype_with_view(self, float_frame, mixed_float_frame):
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by h-vetinari** [comment]
@jreback 
Split up the tests and it's green.

File: pandas/tests/frame/test_dtypes.py [hunk 4]
```diff
@@ -410,41 +409,57 @@ def test_dtypes_gh8722(self):
...
-        mn = self.all_mixed._get_numeric_data().copy()
+        casted = mixed_float_frame.reindex(
+            columns=['A', 'B']).astype('float32')
+        _check_cast(casted, 'float32')
+
+        casted = mixed_float_frame.reindex(
+            columns=['A', 'B']).astype('float16')
+        _check_cast(casted, 'float16')
+
+    def test_astype_mixed_type(self, mixed_type_frame):
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: i don't want the repeated code.

File: pandas/tests/frame/conftest.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/frame/conftest.py]: on 2nd thought, let's see how often this is actually used, so move into the test module

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/conftest.py (hunk 1 above)
Comment (h-vetinari) [reply pandas/tests/frame/conftest.py]: I'll reiterate from [another thread](https://github.com/pandas-dev/pandas/pull/25633#pullrequestreview-212612585), that with these fixturization PRs, I will:
* keep the `TestData` "fixtures" in `tests.frame.conftest`
* any new fixtures will stay in their respective modules
* the effort for #22471 is strictly translation, no breakup of tests
* [reorg of fixtures from `TestData` into submodules to happen later]

File: pandas/tests/frame/test_dtypes.py (hunk 4 above)
Comment (h-vetinari) [reply pandas/tests/frame/test_dtypes.py]: Fine by me, but that means this *should* be done in a separate PR. I'll move the `_check_cast` function to the top of the module.

**Review by h-vetinari — COMMENTED** [review]
Same points as in the other PRs...

**Review by jreback — CHANGES_REQUESTED** [review]
merge master

**Comment by jreback** [comment]
closing as stale

**Comment by h-vetinari** [comment]
@jreback

I had merged master like you asked, this PR is not stale. Please reopen.

**Comment by h-vetinari** [comment]
@jreback @gfyoung @simonjayhawkins
Can you please reopen this PR?

**Comment by h-vetinari** [comment]
Thanks for reopening @simonjayhawkins.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm.

**Comment by jreback** [comment]
thanks @h-vetinari

**Pull request merged**